### PR TITLE
[DON'T MERGE] DATEADD() gives wrong output with DATETIMEOFFSET

### PR DESCRIPTION
**Pull request by bsatarupa**
###Problem description
DATEADD() gives wrong output with DATETIMEOFFSET
Postgres outputs datetimeoffset data by removing/subtracting input timezone value from it.
Fix includes, working on the engine output to include timezone and thereby retain the input timestamp as is.

1> declare @dt datetimeoffset(6)
2> set @dt = '1912-10-25 12:24:32 +10:0'
3> select dateadd(month,1,@dt)
4> go
dateadd                                      
---------------------------------------------
  BABEL output: 1912-11-25 02:24:32.0000000 +10:00

        SQL output: 1912-11-25 12:24:32.0000000 +10:00

###Upgrade
This commit includes adding changes to dateadd_internal_df() to reflect sql behavior.

Test scenarios covered

o Use case based: Original issue
o Null testcases: N/A
o Negative test scenarios: added negative timezone (Permissable range +14:00 to -14:00)
o Boundary conditions, Arbitrary inputs: Testcases including leap year dates, dates across years, dates having positive and negative timezones.
o Testcases calling dateadd function for datetimeoffset from procedures, views and functions
o Expected result correction on existing testcases
o Minor version upgrade tests, Major version upgrade tests : Done
o Performance tests, Tooling impact, Client tests: N/A

Task: BABEL-1625
Signed-off-by: Satarupa Biswas [satarupb@amazon.com](mailto:satarupb@amazon.com)

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.3.0--2.4.0.sql [hunk 1]
```diff
@@ -2505,6 +2505,55 @@ where has_schema_privilege(sch.schema_id, 'USAGE')
 and has_table_privilege(c.oid, 'SELECT,INSERT,UPDATE,DELETE,TRUNCATE,TRIGGER');
 GRANT SELECT ON sys.index_columns TO PUBLIC;
 
+/*
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.3.0--2.4.0.sql]: why is this added to 2.3 to 2.4 in 3 x dev? what about 3.1 to 3.2?

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 2]
```diff
@@ -1222,7 +1222,11 @@ LANGUAGE plpgsql IMMUTABLE;
     but should keep using OPERATOR(sys.+) when input date is in datetimeoffset type.
 */
 CREATE OR REPLACE FUNCTION sys.dateadd_internal_df(IN datepart PG_CATALOG.TEXT, IN num INTEGER, IN startdate datetimeoffset) RETURNS datetimeoffset AS $$
+DECLARE
+	timezone INTEGER;
 BEGIN
+	timezone = sys.babelfish_get_datetimeoffset_tzoffset(startdate)::INTEGER * 2;
```
Comment (surendravishnoi) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql:1228]: this change adds timezone to start date but commit message says opposite, could you please explain

File: contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.3.0--2.4.0.sql (hunk 1 above)
Comment (bsatarupa) [reply contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.3.0--2.4.0.sql]: Added upgarde script from 3.1 to 3.2

File: contrib/babelfishpg_tsql/sql/sys_functions.sql (hunk 2 above)
Comment (bsatarupa) [reply contrib/babelfishpg_tsql/sql/sys_functions.sql:1228]: Modified commit message.

Postgres outputs datetimeoffset data by removing/subtracting input timezone value from it.
Fix includes, working on the engine output to include timezone and thereby retain the input timestamp as is.
1> declare @dt datetimeoffset(6)
2> set @dt = '1912-10-25 12:24:32 +10:0'
3> select dateadd(month,1,@dt)
4> go
dateadd

 BABEL output: 1912-11-25 02:24:32.0000000 +10:00  (after emoving/subtracting input timezone value from it)

   SQL output: 1912-11-25 12:24:32.0000000 +10:00

**Comment by bsatarupa** [comment]
Closing this branch because of some unresolved conflicts.
Please check: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/1340
